---
title: "postUMI_LimmaVoom_DEG_allregions, HOMER_TFEnrichmentAnalysis, LOLAPromoterEnrichment_MicrogliaChIPseq"
author: "Annie Ciernia, Jenn Kim"
date: "`r Sys.Date()`"
output: html_document
---

# load libraries
```{r}
library(ggplot2)
library(dplyr)
library(cowplot)
library(RColorBrewer)
library(gplots)
library(ggrepel)
library(tidyr)
library(edgeR)
library(limma)
library(biomaRt)
library(RColorBrewer)
library(plotrix)
library(marge)
library(gplots)
library(tidyverse)
library(cowplot)
```

# The following code is for analyis of RNASeq in Limma Voom package

## load mm10 genes biomart object and DEGlist object
```{r}
# load mouse mm10 genes biomart object
load("mm10_Ensemble_geneInfo.RData")

# load DEGlist object
load(file="DEGlist.RData")
```

## Filtering and QC plots
```{r}
################################################################################
####################### Before and after filtering reads #######################
################################################################################

# filtering plot
nsamples <- ncol(RGcpm)

colourCount = nsamples
getPalette = colorRampPalette(brewer.pal(9, "Set1"))
fill=getPalette(colourCount)

#plot:
pdf('FilteringCPM_plots.pdf', h=6,w=10)
      par(mfrow=c(1,2))
      
      #prefilter:
      lcpm <- cpm(RG, log=TRUE, prior.count=2)
      
      plot(density(lcpm[,1]), col=fill[1], lwd=2, ylim=c(0,0.5), las=2, 
           main="", xlab="")
      title(main="A. Raw data", xlab="Log-cpm")
      abline(v=0, lty=3)
      for (i in 2:nsamples){
        den <- density(lcpm[,i])
        lines(den$x, den$y, col=fill[i], lwd=2)
      }
      
      #filtered data
      #log-CPM of zero threshold (equivalent to a CPM value of 1) used in the filtering ste
      lcpm <- cpm(RGcpm, log=TRUE, prior.count=2)
      plot(density(lcpm[,1]), col=fill[1], lwd=2, ylim=c(0,0.5), las=2, 
           main="", xlab="")
      title(main="B. Filtered data", xlab="Log-cpm")
      abline(v=0, lty=3)
      for (i in 2:nsamples){
        den <- density(lcpm[,i])
        lines(den$x, den$y, col=fill[i], lwd=2)
      }
dev.off()


################################################################################
################################# Library sizes ################################
################################################################################

#reset library sizes
RGcpm$samples$lib.size <- colSums(RGcpm$counts)

#plot library sizes
pdf('LibrarySizes.pdf',w=30,h=8)
      barplot(RGcpm$samples$lib.size,names=colnames(RGcpm),las=2)
      # Add a title to the plot
      title("Barplot of library sizes")
dev.off()

# Get log2 counts per million
logcounts <- cpm(RGcpm,log=TRUE)

# Check distributions of samples using boxplots
pdf('NonNormalizedLogCPM.pdf',w=30,h=10)
      boxplot(logcounts, xlab="", ylab="Log2 counts per million",las=2)
      # Let's add a blue horizontal line that corresponds to the median logCPM
      abline(h=median(logcounts),col="blue")
      title("Boxplots of logCPMs (unnormalised)")
dev.off()
```

# DE model setup: treatment x brain region interactions, full model
```{r}
# load metadata
info <- read.csv("LPS_DNARNAinfo2.csv")
rownames(info) <- info$Samlple.ID
#info <- info[,c(1:6)]
info <- info %>% dplyr::rename(BrainRegion = Brain.Region)

#make design matrix
info$BrainRegion <- factor(info$BrainRegion)
info$Condition <- factor(info$Condition, levels=c("PBS","1xLPS", "2xLPS", "3xLPS", "4xLPS"))
info$hemisphere <- factor(info$hemisphere)
info$SampleName <- as.character(info$SampleName)
info <- info %>% unite(Condition_BrainRegion, c(Condition, BrainRegion), sep="_", remove=FALSE)
info$Condition_BrainRegion <- factor(info$Condition_BrainRegion)

#set design matrix
design <- model.matrix(~0 + Condition:BrainRegion, data=info) #if using a 0 intercept must set up contrasts
colnames(design) <- make.names(colnames(design))

#Contrasts for conditions:
my.contrasts <- makeContrasts(

  PBSvs1xLPS_FC = Condition1xLPS.BrainRegionPFC - ConditionPBS.BrainRegionPFC,
  PBSvs2xLPS_FC = Condition2xLPS.BrainRegionPFC - ConditionPBS.BrainRegionPFC,
  PBSvs3xLPS_FC = Condition3xLPS.BrainRegionPFC - ConditionPBS.BrainRegionPFC,
  PBSvs4xLPS_FC = Condition4xLPS.BrainRegionPFC - ConditionPBS.BrainRegionPFC,
  
  PBSvs1xLPS_HC = Condition1xLPS.BrainRegionHC - ConditionPBS.BrainRegionHC,
  PBSvs2xLPS_HC = Condition2xLPS.BrainRegionHC - ConditionPBS.BrainRegionHC,
  PBSvs3xLPS_HC = Condition3xLPS.BrainRegionHC - ConditionPBS.BrainRegionHC,
  PBSvs4xLPS_HC = Condition4xLPS.BrainRegionHC - ConditionPBS.BrainRegionHC,
  
  PBSvs1xLPS_STR = Condition1xLPS.BrainRegionSTR - ConditionPBS.BrainRegionSTR,
  PBSvs2xLPS_STR = Condition2xLPS.BrainRegionSTR - ConditionPBS.BrainRegionSTR,
  PBSvs3xLPS_STR = Condition3xLPS.BrainRegionSTR - ConditionPBS.BrainRegionSTR,
  PBSvs4xLPS_STR = Condition4xLPS.BrainRegionSTR - ConditionPBS.BrainRegionSTR,
  
levels=colnames(design))
```

## Normalization TMM
```{r}
DGE=calcNormFactors(RGcpm,method =c("TMM")) #TMM normalization for library size/composition

pdf('VoomTrend.pdf',w=6,h=4)
    v=voom(DGE,design,plot=T)
dev.off()
```

## Voom
```{r}
# SampleName as random effects (repeated measure): not included in the model
# after running below code, determined that we should not include SampleName in final fit (negative consensus correlation)
cor <- duplicateCorrelation(v, design, block=info$SampleName)
cor$consensus.correlation # -0.003924953 = very small correlation value. negative values should be removed, resume with usual modeling approach of accounting for GROUP fixed effect only in design matrix. 

fit <- lmFit(v, design)

fit2 <- contrasts.fit(fit,my.contrasts)
fit2 <- eBayes(fit2,robust=TRUE )

# Voom trend
pdf('PlotSA_VoomTrend.pdf',w=6,h=4)
    plotSA(fit2, main="Final model: Meanâˆ’variance trend",ylim=c(0,4))
dev.off()

#box plots for the voom normalised data to compare to before normalisation (only RLE)
#v$E already log2CPM
par(mfrow=c(1,2))
boxplot(logcounts, xlab="", ylab="Log2 counts per million",las=2,main="unnormalised logCPM")
## Let's add a blue horizontal line that corresponds to the median logCPM
abline(h=median(logcounts),col="blue")
boxplot(v$E, xlab="", ylab="Log2 counts per million",las=2,main="TMM & Voom transformed logCPM")
## Let's add a blue horizontal line that corresponds to the median logCPM
abline(h=median(v$E),col="blue")

#plot only normalized data
pdf('TMM_VOOM_NormalizedLogCPM.pdf',w=30,h=10)
    boxplot(v$E, xlab="", ylab="Log2 counts per million",las=2,main="TMM & Voom transformed logCPM")
    # Let's add a blue horizontal line that corresponds to the median logCPM
    abline(h=median(logcounts),col="blue")
dev.off()
```

## MDS plots
```{r}
# mds plot colored by brain region
pdf(file = "MDSplot_Region.pdf", wi = 8, he = 8, useDingbats=F)
    #plot MDS for condition
    levels(info$BrainRegion)
    col.cell <- c("skyblue","purple","orange")[info$BrainRegion]
    data.frame(info$BrainRegion,col.cell)
    
    # MDS with group colouring
    #par(xpd = T, mar = par()$mar + c(0,0,0,7))
    plotMDS(v,col=col.cell)
    # Legend
    legend("topleft",
           fill=c("skyblue","purple","orange"),
           legend=levels(info$BrainRegion),
           cex = 0.8)
    # Add a title
    title("Brain Region MDS Plot")
dev.off()


# mds plot colored by treatment
pdf(file = "MDSplot_Treatment.pdf", wi = 16, he = 8, useDingbats=F)
    #plot MDS for condition
    levels(info$Condition)
    col.cell <- c("skyblue","purple","orange","green","blue")[info$Condition]
    data.frame(info$Condition,col.cell)
    
    par(mfrow=c(1,2))
    
    # MDS with group colouring
    plotMDS(v,dim=c(3,4),col=col.cell)
    # Legend
    legend("topright",
           fill=c("skyblue","purple","orange","green","blue"),
           legend=levels(info$Condition),
           cex = 0.8)
    # Add a title
    title("Treatment Condition MDS Plot")
    
    plotMDS(v,dim=c(3,5),col=col.cell)
    # Legend
    legend("topright",
           fill=c("skyblue","purple","orange","green","blue"),
           legend=levels(info$Condition),
           cex = 0.8)
    # Add a title
    title("Treatment Condition MDS Plot")
dev.off()


#mds plot colored by Hemisphere:
pdf(file = "MDSplot_Hemisphere.pdf", wi = 8, he = 8, useDingbats=F)
    levels(info$hemisphere)
    col.cell <- c("orange","green")[info$hemisphere]
    data.frame(info$hemisphere,col.cell)
    
    # MDS with group colouring
    plotMDS(v,col=col.cell)
    # Legend
    legend("topleft",
           fill=c("orange","green"),
           legend=levels(info$hemisphere),
           cex = 0.8)
    # Add a title
    title("Hemisphere MDS Plot")
dev.off()
```

## Average CPM for each condition
```{r}
#get log2CPM counts from voom and put in dataframe:
countdf <- as.data.frame(v$E)
countdf$GeneID<- rownames(v$E)

#summarize 
countdf2 <- countdf %>% group_by(GeneID) %>% gather(sample,log2CPM, 1:(ncol(countdf)-1)) 
countdf2 <- as.data.frame(countdf2)
countdf3 <-merge(countdf2,info,by.x="sample",by.y="Samlple.ID")

# by combo of brains + treatment
GeneSummary <- countdf3 %>% group_by(GeneID, Condition_BrainRegion) %>% 
  summarize(meanlog2CPM = mean(log2CPM)) %>%
  ungroup()  %>%
 # dplyr::select(GeneID,Group,meanlog2CPM) %>%
  spread(Condition_BrainRegion, meanlog2CPM)

#add in gene symbol information
geneinfo <- v$genes
GeneSummary2 <- merge(GeneSummary, geneinfo ,by.x="GeneID",by.y="ensembl_gene_id")

#go back to CPM from log2: 2^log2CPM = CPM
GeneSummaryCPM <- countdf3 %>% group_by(GeneID, Condition_BrainRegion) %>% 
  summarize(meanlog2CPM = mean(log2CPM)) %>%
  ungroup()  %>%
  mutate(meanCPM = 2^meanlog2CPM) %>%
  dplyr::select(GeneID, Condition_BrainRegion, meanCPM) %>%
  spread(Condition_BrainRegion, meanCPM)

#add in gene symbol information
GeneSummaryCPM2 <- merge(GeneSummaryCPM,geneinfo,by.x="GeneID",by.y="ensembl_gene_id")

save(GeneSummary2,GeneSummaryCPM2,file="AverageCPMandLog2CPM.RData")
```

## DE analysis
```{r}
str(fit2)
#summary(decideTests(fit2,adjust.method="fdr", method="global"))
dt <- decideTests(fit2,adjust.method="BH", method="separate")
sum <- summary(dt)
write.csv(sum,"SummaryCount_DEGs.csv")

dt <- as.data.frame(dt)
tfit <- fit2

#######################################################################
#get out DE lists for each contrast:
#######################################################################
library("calibrate")
comparisons=(coef(tfit))
comparisons=colnames(comparisons)

comp_out <- as.data.frame(rownames(v$E))
names(comp_out) <- c("GeneID")

SumTableOut <- NULL

for(i in 1:length(comparisons)){
  #comparison name
  comp=comparisons[i]
  print(comp)
  #make comparisons 
  
  tmp=topTreat(tfit,coef=i,number=nrow(comp_out),adjust.method="BH")
  dim(tmp[(tmp$adj.P.Val<0.05),]) # number of DE genes
  
  #LogFC values:https://support.bioconductor.org/p/82478/
  tmp$direction <- c("none")
  tmp$direction[which(tmp$logFC > 0)] = c("Increase")
  tmp$direction[which(tmp$logFC < 0)] = c("Decrease")
  
  tmp$significance <- c("nonDE")
  tmp$significance[which(tmp$adj.P.Val <0.05)] = c("DE")
  
  #summary counts table based on Ensemble Gene ID counts:
  SumTable <- table(tmp$significance,tmp$direction)
  SumTable <- as.data.frame(SumTable)
  SumTable$comparison <- paste(comp)
  SumTableOut <- rbind(SumTable,SumTableOut)
  
  #get geneids  
  tmp$GeneID <- rownames(tmp)
  
  #gene gene names and expression levels
  tmp2 <- tmp
  
  tmp2$comparison <- paste(comp)
  
  #add in CPM values
  GeneSummaryCPM3 <- GeneSummaryCPM2[,c(1:21)]
  DECPM <- merge(tmp2, GeneSummaryCPM3,by="GeneID")
  
  write.csv(DECPM,file = paste(comp,"_DEgenes.csv"))
  
  #save to output:
  #merge <- merge(comp_out,tmp2, by= "GeneID")
  merge2 <- tmp2 %>% dplyr::select(GeneID,logFC,t,P.Value,adj.P.Val,direction,significance)
  colnames(merge2) <- paste(colnames(merge2),comp,sep=".")
  colnames(merge2)[1] <- c("GeneID")
  comp_out <- merge(comp_out, merge2, by="GeneID")
  
  #data for plot with gene names:
  genenames <- tmp2 %>% dplyr::select(adj.P.Val,logFC,GeneID,mgi_symbol) %>% distinct()
  
  #names for plots
  plotname <- gsub("\\."," ",comp)
  plotname <- gsub("vs"," vs ",plotname)
  
  #volcano plot
  pdf(file = paste(comp,"_Volcano.pdf", sep=""), wi = 9, he = 6, useDingbats=F)
  
  with(genenames, plot(logFC, -log10(adj.P.Val), pch=20,col="gray", main=paste(plotname,"\nVolcano plot", sep=" "), ylab =c("-log10(adj.pvalue)"),xlab =c("Log Fold Change") ))
  
  #color points red when sig and log2 FC > 2 and blue if log2 FC < -2 
  with(subset(genenames, logFC < -2 & -log10(adj.P.Val) > -log10(.05)), points(logFC, -log10(adj.P.Val), pch=20, col="blue"))
  with(subset(genenames, logFC > 2 & -log10(adj.P.Val) > -log10(.05)), points(logFC, -log10(adj.P.Val), pch=20, col="red"))
  
  #add lines
  abline(h = -log10(.05), col = c("black"), lty = 2, lwd = 1)
  abline(v = c(-2,2), col = "black", lty = 2, lwd = 1)
  
  # Label points with the textxy function from the calibrate plot
  library(calibrate)
  with(subset(genenames, adj.P.Val<0.05 & abs(logFC)>2), textxy(logFC, -log10(adj.P.Val), labs=mgi_symbol, cex=.8))
  
  dev.off()
  
}


write.csv(SumTableOut,"SummaryTableDEgenes.csv")

#master outfile
mout <- merge(comp_out, GeneSummaryCPM2, by="GeneID")
write.csv(mout,"AllDEG_AllConditions.csv")
```

# DEG heatmaps
```{r}
heatmapgenes <- mout %>% filter(significance.PBSvs1xLPS_FC == "DE"|
                                significance.PBSvs2xLPS_FC == "DE"|
                                significance.PBSvs3xLPS_FC == "DE"|
                                significance.PBSvs4xLPS_FC == "DE"|
                                  
                                significance.PBSvs1xLPS_HC == "DE"|
                                significance.PBSvs2xLPS_HC == "DE"|
                                significance.PBSvs3xLPS_HC == "DE"|
                                significance.PBSvs4xLPS_HC == "DE"|
                                  
                                significance.PBSvs1xLPS_STR == "DE"|
                                significance.PBSvs2xLPS_STR == "DE"|
                                significance.PBSvs3xLPS_STR == "DE"|
                                significance.PBSvs4xLPS_STR == "DE") 
nrow(heatmapgenes)

##############################################
##### heatmap with biological replicates #####
##############################################

# for heatmaps:
heatmapgenes <- heatmapgenes$GeneID 

# v$E is log2CPM
# cpm data
log2cpm <- v$E
# matrix of gene expression fold changes
data <- log2cpm %>% as.data.frame()
data2 <- data %>% filter(rownames(data) %in% heatmapgenes)
mat <- data.matrix(data2)
colnames(mat) <- gsub("AC","",colnames(mat))
mat <- mat[,c(21,32,34:40,22:31,33,1:20,41:60)]

# if you want to look at CPM, scaled but not centered within each brain region
mat <- t(2^mat)
mat_FC <- scale(mat[1:20,], center=FALSE)
mat_HC <- scale(mat[21:40,], center=FALSE)
mat_STR <- scale(mat[41:60,], center=FALSE)
mat <- rbind(mat_FC, mat_HC , mat_STR)
mat <- t(mat)

# metadata
info <- read.csv("LPS_DNARNAinfo2.csv")
rownames(info) <- info$Samlple.ID
info$sample <- as.character(info$Samlple.ID)
info$sample <- gsub("AC","",info$sample)
info <- info %>% dplyr::select(sample,SampleName,Condition,Brain.Region)

# annotation columns
anno_col <- as.data.frame(info[,3:4])
rownames(anno_col) <- gsub("AC","",rownames(anno_col))
anno_col$Condition <- factor(anno_col$Condition, levels=c("PBS","1xLPS","2xLPS","3xLPS","4xLPS"))
anno_col$Brain.Region <- factor(anno_col$Brain.Region, levels=c("PFC","STR","HC"))

# quantile breaks
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

dat <- data.frame(values = as.numeric(mat))
mat_breaks <- quantile_breaks(mat, n = 11)

dat_colors <- data.frame(
  xmin = mat_breaks[1:(length(mat_breaks)-1)],
  xmax = mat_breaks[2:length(mat_breaks)],
  ymin = 0,
  ymax = max(density(mat, bw = "SJ")$y),
  fill = rev(inferno(length(mat_breaks) - 1)),
  stringsAsFactors = FALSE
)

# quantile breaks for legend
pdf("Heatmap_SigDEGs_LegendBreakDistribution.pdf",height=2,width = 5)
ggplot() +
  geom_rect(
    data = dat_colors,
    mapping = aes(
      xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax, fill = fill
    )
  ) +
  geom_density(
    data = dat,
    mapping = aes(values),
    bw = "SJ", color = "cyan"
  ) +
  scale_fill_manual(values = dat_colors$fill) +
  theme(legend.position = "none") +
  labs(title = "Quantile breaks")
dev.off()

# number of data per break
pdf("Heatmap_SigDEGs_LegendBreakCounts.pdf",height=10,width = 10)
dat2 <- as.data.frame(table(cut(
  mat, mat_breaks
)))
dat2$fill <- inferno(nrow(dat2))
ggplot() +
  geom_bar(
    data = dat2,
    mapping = aes(x = Var1, weight = Freq, fill = Var1),
    color = "black", size = 0.1
  ) +
  coord_flip() +
  scale_fill_manual(values = dat2$fill) +
  theme(legend.position = "none") +
  labs(y = "data points", x = "breaks",
       title = "Number of data points per color")
dev.off()

library(pheatmap)

# cluster and extract labels for row annotations
set.seed(1)
hm <- pheatmap(mat, border_color=F, annotation_col=anno_col, 
         cluster_rows=T, cluster_cols=F, scale="row",
         show_rownames=F, gaps_col=c(20,40),
         color=inferno(length(mat_breaks)-1),
         breaks=mat_breaks)

# get out cluster row names
clusters <- data.frame(cutree(hm$tree_row, k=4)) %>% rename(clusterID=cutree.hm.tree_row..k...4.)
# update cluster labels
clusters <- clusters %>% mutate(clusterID = 
                                  case_when(clusterID=="1" ~ "2xLPS sensitive",
                                            clusterID=="2" ~ "4xLPS sensitive",
                                            clusterID=="3" ~ "LPS-decreased-1",
                                            clusterID=="4" ~ "LPS-decreased-2"))
clusters$clusterID <- as.factor(clusters$clusterID)

# cpm scaled within brain region, then scaled by row (gene)
pdf("Heatmap_SigDEGs.pdf",height=10,width = 10)
pheatmap(mat, border_color=F, annotation_col=anno_col, 
               cluster_rows=T, cluster_cols=F, cutree_rows=4, scale="row", annotation_row=clusters,
               show_rownames=F, gaps_col=c(20,40),
               color=inferno(length(mat_breaks)-1), angle_col = 270,
               breaks=mat_breaks)
dev.off()
```

# DEG venn diagrams, by condition and by brain region
```{r}
library(VennDiagram)
#venn for LPS treatments:
#The number of genes that are not DE in either comparison are marked in the bottom-right.

colnames(dt)

# 1xLPS>PBS
pdf("Venn_1xLPS-greater-PBS.pdf",height=10,width = 10)
vennDiagram(dt[,c(1,5,9)],
            circle.col=c("turquoise","purple","salmon","green"), include="up")
dev.off()

# 1xLPS<PBS
pdf("Venn_1xLPS-lessthan-PBS.pdf",height=10,width = 10)
vennDiagram(dt[,c(1,5,9)],
            circle.col=c("turquoise","purple","salmon","green"), include="down")
dev.off()

# 2xLPS>PBS
pdf("Venn_2xLPS-greater-PBS.pdf",height=10,width = 10)
vennDiagram(dt[,c(2,6,10)],
            circle.col=c("turquoise","purple","salmon","green"), include="up")
dev.off()

# 2xLPS<PBS
pdf("Venn_2xLPS-lessthan-PBS.pdf",height=10,width = 10)
vennDiagram(dt[,c(2,6,10)],
            circle.col=c("turquoise","purple","salmon","green"), include="down")
dev.off()

# 3xLPS>PBS
pdf("Venn_3xLPS-greater-PBS.pdf",height=10,width = 10)
vennDiagram(dt[,c(3,7,11)],
            circle.col=c("turquoise","purple","salmon","green"), include="up")
dev.off()

# 3xLPS<PBS
pdf("Venn_3xLPS-lessthan-PBS.pdf",height=10,width = 10)
vennDiagram(dt[,c(3,7,11)],
            circle.col=c("turquoise","purple","salmon","green"), include="down")
dev.off()

# 4xLPS>PBS
pdf("Venn_4xLPS-greater-PBS.pdf",height=10,width = 10)
vennDiagram(dt[,c(4,8,12)],
            circle.col=c("turquoise","purple","salmon","green"), include="up")
dev.off()

# 4xLPS<PBS
pdf("Venn_4xLPS-lessthan-PBS.pdf",height=10,width = 10)
vennDiagram(dt[,c(4,8,12)],
            circle.col=c("turquoise","purple","salmon","green"), include="down")
dev.off()


# FrontalCortex Treatment-specific genes
pdf("Venn_FC.pdf",height=10,width = 10)
vennDiagram(dt[,c(1:4)],
            circle.col=c("turquoise","purple","salmon","green"), include="up")
dev.off()

# Hippocampus Treatment-specific genes
pdf("Venn_HC.pdf",height=10,width = 10)
vennDiagram(dt[,c(5:8)],
            circle.col=c("turquoise","purple","salmon","green"), include="up")
dev.off()

# Striatum Treatment-specific genes
pdf("Venn_STR.pdf",height=10,width = 10)
vennDiagram(dt[,c(9:12)],
            circle.col=c("turquoise","purple","salmon","green"), include="up")
dev.off()
```

# GO enrichment of each cluster (2xLPS, 4xLPS, LPSdecreased-1 and 2) using ClusterCompare
```{r}
library("clusterProfiler")
library(org.Mm.eg.db) #mouse

# get ensembl gene IDs of the clusters you are interested in
factor(clusters$clusterID) %>% levels()
x = list("2xLPS" = clusters %>% filter(clusterID=="2xLPS sensitive") %>% rownames(),
         "4xLPS" = clusters %>% filter(clusterID=="4xLPS sensitive") %>% rownames(),
         "LPSDecreased" = c(clusters %>% filter(clusterID=="LPS-decreased-1") %>% rownames(), 
                            clusters %>% filter(clusterID=="LPS-decreased-2") %>% rownames()))

# save lists
clusters %>% filter(clusterID=="2xLPS sensitive") %>% rownames() %>% write("2xLPS_cluster_genes.txt")
clusters %>% filter(clusterID=="4xLPS sensitive") %>% rownames() %>% write("4xLPS_cluster_genes.txt")
c(clusters %>% filter(clusterID=="Downregulated with LPS-1") %>% rownames(),
  clusters %>% filter(clusterID=="Downregulated with LPS-2") %>% rownames()) %>% write("LPSDecreased_cluster_genes.txt")

library(purrr) # remove NAs
y = list("2xLPS" = GeneSummaryCPM2 %>% filter(GeneSummaryCPM2$GeneID %in% x$`2xLPS`) %>% dplyr::pull(entrezgene_id) %>% discard(is.na),
         "4xLPS" = GeneSummaryCPM2 %>% filter(GeneSummaryCPM2$GeneID %in% x$`4xLPS`) %>% dplyr::pull(entrezgene_id) %>% discard(is.na),
         "LPSDecreased" = GeneSummaryCPM2 %>% filter(GeneSummaryCPM2$GeneID %in% x$LPSDownregulated) %>% dplyr::pull(entrezgene_id) %>% discard(is.na))

# compareCluster: given a list of gene sets, this function will compute profiles of each gene cluster
CompareGO <- compareCluster(geneClusters=y, pvalueCutoff=0.01, qvalueCutoff=0.05, 
                            readable=T, pAdjustMethod="BH", OrgDb="org.Mm.eg.db", 
                            fun="enrichGO", ont="BP")
      
GO_sum <- as.data.frame(CompareGO)
View(GO_sum)
write.csv(GO_sum, "GOclusters.csv")

# plots
pdf("GOcluster_dotplot.pdf", height=8, width=7)
dotplot(CompareGO)
dev.off()

pdf("GOclusters_cnetplot.pdf",height=9,width = 16)
cnetplot(CompareGO)
dev.off()
```

# The following code is for Homer transcription factor analysis and integration with 2xLPS cluster DEGs

# load in known results for 2xLPS cluster TF enrichments
```{r}
library(marge)
path_HomerResults = "2xLPS_cluster_genes_output/" # Supplementary file S1
known <- read_known_results(path = path_HomerResults, homer_dir = TRUE)

#filter for signficant qvalue
motifsSig <- known %>% filter(fdr < 0.05)
motifsSig$DEGCluster <- "2xLPS-sensitive"
```

# PLOTS

# background vs. target % for known Homer motifs
```{r}
order <- motifsSig %>% arrange(desc(fdr))
order <- order$motif_name

motifsSig2 <- motifsSig %>% arrange(desc(fdr)) %>%
  dplyr::select(DEGCluster, motif_name, motif_family, tgt_pct, bgd_pct) %>%
  gather(key=Percentwithmotif,value=percent, 4:5) %>%
  mutate(percent=percent*100) %>%
  mutate(Percentwithmotif = 
           case_when(Percentwithmotif=="tgt_pct" ~ "Target %",
                     Percentwithmotif=="bgd_pct" ~ "Background %")) %>%
  mutate(motif_family = 
           case_when(motif_family=="bZIP,IRF" ~ "IRF",
                     motif_family=="ETS:IRF" ~ "IRF",
                     TRUE ~ motif_family))

motifsSig2 <- as.data.frame(motifsSig2)
motifsSig2$Percentwithmotif <- factor(motifsSig2$Percentwithmotif, levels=c("Target %", "Background %"))
motifsSig2$motif_name <- factor(motifsSig2$motif_name, levels=order)

pdf("Homer_SigMotif.dotPlot.pdf", width=12.5, height=3)
ggplot(motifsSig2, aes(y = motif_name, x = DEGCluster, group=Percentwithmotif)) +
  facet_wrap(~motif_family, scales = "free_y", ncol=5) +
  theme_bw() +
  geom_point( alpha=1, aes(size = percent, color=Percentwithmotif),position=position_dodge(width = .2)) + 
  scale_color_manual(values = c("Target %" = "blue", "Background %" = "darkgrey")) +
  xlab("Transcription factor family") +
  ylab("Transcription factor motif") + 
  labs(color="% with motif", size="Percentage") +
  ggtitle("Transcription factor motif enrichment of 2xLPS-sensitive cluster DEGs") +
  guides(x="none")
dev.off()
```

# heatmaps of known Homer motif annotations in 2xLPS DEG promoters for IRF family transcription factors
# continuation of "# DEG heatmaps" section above -- adding TF motif enrichments as row annotations
```{r}
############################################################
#read in annotation of homer motifs in individual promoters
##########################################################
motifs <- read.delim("2xLPS_cluster_genes_outputmotifs.txt") # Supplementary File S1
motifs$DEGCluster <- "2xLPS-sensitive"

#clean up and fix motif names and families manually
motifs$motiffamily <- motifs$Motif.Name

tmp <- motifs %>% 
  separate(motiffamily, into=c("keep","trash1","trash2"), sep="/", remove=TRUE) %>%
  dplyr::select(-c("trash1","trash2")) 

tmp$keep <- gsub("[(]","_",tmp$keep)
tmp$keep <- gsub("[)]","",tmp$keep)

motifs2 <- tmp %>% 
  separate(keep, into=c("motif_name", "motif_family"), sep="_", remove=TRUE)

write.csv(motifs2,"PromoterLocations.knownHomerMotifs.csv")

##############################
#heatmap of IRF family genes
##############################

# remove IRF4 and Stat3: didn't reach qvalue significance, and Fos: wasn't included in Homer's known output so wasn't plotted in figureA/B
remove <- c("IRF4","Stat3","Fos")
IRFgenes <- motifs2 %>% filter(!(motif_name %in% remove))

# add binary for yes, TF motif present or no, TF not present
IRFgenes2 <- IRFgenes
unique_motifnames <- unique(IRFgenes2$motif_name)
for(m in unique_motifnames){
  IRFgenes2 <- IRFgenes2 %>%
    mutate(
      !!paste0(m) := ifelse(motif_name==m, 1, 0)
    )
}

# group by individual gene and collapse
targets <- IRFgenes2 %>%
  group_by(Ensembl) %>% 
  summarise(across("ISRE":"NFkB-p65", sum))

# merge heatmap annotations with new TF-specific columns, convert any # >1 to one (since it's a binarized yes/no presence)
rowanno <- as.data.frame(clusters) %>% rownames_to_column(var="Ensembl")
rowanno2 <- merge(rowanno, targets, by="Ensembl", all=TRUE)
rowanno2[is.na(rowanno2)] <- 0
rowanno2 <- rowanno2 %>% mutate(across("ISRE":"NFkB-p65", ~ifelse(.>1, 1, .)))
rowanno2 <- rowanno2 %>% column_to_rownames(var="Ensembl")
rowanno2 <- rowanno2 %>%
  mutate(across("ISRE":"NFkB-p65", as.character))

# filter for just the 2xLPS cluster
targets2xLPS <- rowanno2 %>% filter(clusterID=="2xLPS sensitive")
mat_motif <- mat[rownames(mat) %in% rownames(targets2xLPS),]

# remove clusterID annotation since there is only one cluster
rowanno2 <- rowanno2[,c(2:ncol(rowanno2))]

# order by family
rowanno3 <- rowanno2[c(12,15,10,2,22,20,13,9,1,11,16:19,14,8,7,4,5,6,3,21)]

# heatmap
pdf("Heatmap_2xLPSClusterDEGs_TFmotifs_ALL.pdf",height=7,width = 18)
pheatmap(mat_motif, border_color=F, annotation_col=anno_col, treeheight_row=0,
               cluster_rows=T, cluster_cols=F, scale="row", 
               annotation_row=rowanno3,
               annotation_colors=list(ISRE=c("1"="black","0"="white"),
                                      STAT1=c("1"="black","0"="white"),
                                      `AP-1`=c("1"="black","0"="white"),
                                      Fosl2=c("1"="black","0"="white"),
                                      Atf3=c("1"="black","0"="white"),
                                      `Jun-AP1`=c("1"="black","0"="white"),
                                      Fra1=c("1"="black","0"="white"),
                                      JunB=c("1"="black","0"="white"),
                                      T1ISRE=c("1"="black","0"="white"),
                                      `Stat3+il21`=c("1"="black","0"="white"),
                                      IRF1=c("1"="black","0"="white"),
                                      PRDM1=c("1"="black","0"="white"),
                                      `bZIP:IRF`=c("1"="black","0"="white"),
                                      BATF=c("1"="black","0"="white"),
                                      STAT4=c("1"="black","0"="white"),
                                      IRF2=c("1"="black","0"="white"),
                                      IRF3=c("1"="black","0"="white"),
                                      IRF8=c("1"="black","0"="white"),
                                      `PU.1:IRF8`=c("1"="black","0"="white"),
                                      `NFkB-p65-Rel`=c("1"="black","0"="white"),
                                      Fra2=c("1"="black","0"="white"),
                                      `NFkB-p65`=c("1"="black","0"="white")),
               show_rownames=F, gaps_col=c(20,40),
               color=inferno(length(mat_breaks)-1), angle_col = 270,
               breaks=mat_breaks)
dev.off()
```

# The following code is for LOLA promoter enrichment for microglia ChIP-seq peaks

# generate .bed files of 2xLPS cluster DEGs
```{r}
#read in all mm10 gene promoters
allmm10 <- read.table("mm10ensembl_Prom_1kb.2kb.bed", header=F)
colnames(allmm10) <- c("chr","start","end","ensembl_gene_id","blank","strand")

#only keep 2xLPS-sensitive cluster gene promoter information
mm10_2xLPS <- allmm10 %>% filter(ensembl_gene_id %in% rownames(targets2xLPS))

write.table(mm10_2xLPS, file="2xLPScluster_DEGPromoters.bed", sep="\t",col.names=F, row.names=F, quote=F)

#sub 0 and 1 for annotate peak files
mm10_2xLPS$strand <- gsub("\\-","1",mm10_2xLPS$strand)
mm10_2xLPS$strand <- gsub("\\+","0",mm10_2xLPS$strand)
mm10_2xLPS$chr <- gsub("chr","",mm10_2xLPS$chr)

write.table(mm10_2xLPS, file="2xLPScluster_DEGPromotersStrand.bed", sep="\t",col.names=F, row.names=F, quote=F)
```

# generate .bed files of background (all genes expressed in experiment)
```{r}
#read in all mm10 gene promoters
allmm10 <- read.table("mm10ensembl_Prom_1kb.2kb.bed", header=F)
colnames(allmm10) <- c("chr","start","end","ensembl_gene_id","blank","strand")

# only keep genes expressed in experiments
bkgd <- read.csv("HOMER_Background_AllGenes.txt", header=F)
bkgd_genes <- bkgd$V1
mm10_background <- allmm10 %>% filter(ensembl_gene_id %in% bkgd_genes)
getwd()

write.table(mm10_background, file="BackgroundAllExpressedGenes_DEGPromoters.bed", sep="\t",col.names=F, row.names=F, quote=F)

#sub 0 and 1 for annotate peak files
mm10_background$strand <- gsub("\\-","1",mm10_background$strand)
mm10_background$strand <- gsub("\\+","0",mm10_background$strand)
mm10_background$chr <- gsub("chr","",mm10_background$chr)

write.table(mm10_background, file="BackgroundAllExpressedGenes_DEGPromotersStrand.bed", sep="\t",col.names=F, row.names=F, quote=F)
```

# lola
```{r}
##############################################################################################################
#lola
##############################################################################################################
library(LOLA)

setwd("C:/Users/Analysis2 Computer/Desktop/Jenn/LOLAPromoterEnrichment/")

#read in 2xLPS cluster bed file:
dat <- readBed(file = "2xLPScluster_DEGPromoters_SORTED.bed")

#load all background promoters
Background <- readBed(file = "BackgroundAllExpressedGenes_DEGPromoters_SORTED.bed")

regionDB <- loadRegionDB(dbLocation = "Collections/", # can be found as .zip file on Github repo
                         useCache = F, limit = NULL)

#one tailed fisher exact test:
Results <- runLOLA(userSets = dat, userUniverse = Background, regionDB = regionDB, minOverlap = 1, cores=4, redefineUserSets = FALSE)
writeCombinedEnrichment(combinedResults = Results, outFolder = "LOLAPromoterOverlaps", includeSplits=F)

########################################################################################################################
merge <- Results
names(merge)[names(merge) == 'support'] <- 'userSet.in.target.list'
names(merge)[names(merge) == 'b'] <- 'NonuserSet.in.target.list'
names(merge)[names(merge) == 'c'] <- 'userSet.not.in.target.list'
names(merge)[names(merge) == 'd'] <- 'NonuserSet.not.in.target.list'

#% enrichment
merge$percent_userSet_in_Target <- (merge$userSet.in.target.list/(merge$userSet.in.target.list + merge$userSet.not.in.target.list)*100)
merge$percent_BG_in_Target <- (merge$NonuserSet.in.target.list/(merge$NonuserSet.in.target.list + merge$NonuserSet.not.in.target.list)*100)

#fold enrichment relative to background
merge$FC <- (merge$percent_userSet_in_Target - merge$percent_BG_in_Target)/merge$percent_BG_in_Target  

write.csv(merge,"OneSidedFisherExact_AllEnrichements.csv")

enrichments <- merge

# promoter overlap plots
plot2 <-  enrichments %>% 
  arrange(collection, description)

order <- plot2$description
order <- order[c(1:4,7,5,6,8:16)] # fix so that IRF8 WT (Saeki 2023) shows up in chronological order by timepoint

plot3 <- plot2 %>% 
  gather(key=Percentwithpeak, value=percent, 24:25) %>%
  mutate(Percentwithpeak = 
           case_when(Percentwithpeak == "percent_userSet_in_Target" ~ "Target %",
                     Percentwithpeak == "percent_BG_in_Target" ~ "Background %"))

plot3 <- as.data.frame(plot3)
plot3$Percentwithpeak <- factor(plot3$Percentwithpeak, levels=c("Target %", "Background %"))
plot3$description <- factor(plot3$description, levels=order)
plot3$DEGCluster <- "2xLPS-sensitive"

plot3$Percentwithpeak[plot3$qValue >=.05] <- NA

pdf("LOLA_SigPeaks.dotPlot.pdf", width=6.9, height=4.25)
ggplot(plot3, aes(y = description, x = DEGCluster, group=Percentwithpeak)) +
  #theme_classic() +
  theme_bw() +
  geom_point(alpha=1, aes(size = percent, color=Percentwithpeak), position=position_dodge(width = .7)) + 
  scale_color_manual(values = c("Target %" = "darkred", "Background %" = "darkgrey"), na.value=aes(alpha=0.1)) +
  xlab("2xLPS-sensitive cluster") +
  ylab("Dataset description") + 
  labs(color="% with peak", size="Percentage") +
  #ggtitle("ChIP-seq peak enrichment of 2xLPS-sensitive cluster DEGs") +
  guides(x="none")
dev.off()
```


